In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy import stats 


In [ ]:
train=pd.read_csv('Train.csv')
test=pd.read_csv('Test.csv')


In [ ]:
#previwe of the data
train

ID  country_code  region   age  ...  FQ33  FQ34  FQ37  Target
0       ID_000J8GTZ             1       6  35.0  ...   1.0   1.0     0       0
1       ID_000QLXZM            32       7  70.0  ...   1.0   2.0     0       0
2       ID_001728I2            71       7  22.0  ...   2.0   1.0     1       0
3       ID_001R7IDN            48       3  27.0  ...   1.0   1.0     1       0
4       ID_0029QKF8            25       0  79.0  ...   1.0   1.0     1       0
...             ...           ...     ...   ...  ...   ...   ...   ...     ...
108441  ID_ZZYY7RTO           111       4  23.0  ...   1.0   NaN     1       0
108442  ID_ZZYZTTC6            77       4  60.0  ...   1.0   1.0     1       0
108443  ID_ZZZ3OW3S            42       2  59.0  ...   1.0   2.0     1       1
108444  ID_ZZZLDXE8            57       7  79.0  ...   1.0   1.0     1       0
108445  ID_ZZZMYW1F           110       2  74.0  ...   1.0   NaN     1       1

[108446 rows x 42 columns]

In [ ]:
test

ID  country_code  region   age  ...  FQ32  FQ33  FQ34  FQ37
0      ID_000YI58E            39       2  22.0  ...   2.0   1.0   1.0     0
1      ID_001SP4JF            30       2  62.0  ...   1.0   1.0   1.0     0
2      ID_001VOF6S            65       4  35.0  ...   NaN   1.0   NaN     0
3      ID_0030LULG           123       0  24.0  ...   NaN   1.0   1.0     1
4      ID_0037PZ3R            67       2  25.0  ...   NaN   2.0   1.0     1
...            ...           ...     ...   ...  ...   ...   ...   ...   ...
46472  ID_ZZYOTVBJ           112       3  38.0  ...   NaN   1.0   1.0     1
46473  ID_ZZYSX122            82       2  77.0  ...   NaN   1.0   2.0     1
46474  ID_ZZYXQDSD            93       5  26.0  ...   NaN   1.0   1.0     1
46475  ID_ZZZH9SS4             6       7  59.0  ...   NaN   1.0   1.0     1
46476  ID_ZZZTLL7U           105       0  24.0  ...   NaN   1.0   1.0     1

[46477 rows x 41 columns]

**we do see a lot of null values so seems we have a lot of thinking do to**

# Univariate Analysis 

## Whole dataset

In [ ]:
#lets get a look at what the dataset contains the different data types and number of features
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108446 entries, 0 to 108445
Data columns (total 42 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   ID            108446 non-null  object 
 1   country_code  108446 non-null  int64  
 2   region        108446 non-null  int64  
 3   age           108124 non-null  float64
 4   FQ1           108446 non-null  int64  
 5   FQ2           49124 non-null   float64
 6   FQ3           46218 non-null   float64
 7   FQ4           108446 non-null  int64  
 8   FQ5           21185 non-null   float64
 9   FQ6           60659 non-null   float64
 10  FQ7           60620 non-null   float64
 11  FQ8           108446 non-null  int64  
 12  FQ9           108446 non-null  int64  
 13  FQ10          108446 non-null  int64  
 14  FQ11          83876 non-null   float64
 15  FQ12          108446 non-null  int64  
 16  FQ13          108446 non-null  int64  
 17  FQ14          108446 non-null  int64  
 18  FQ15

we have 41 features and one target variable.The number of rows is *108446* in the training dataset we do see lots of misisng values and a mixture of categorical and numerical values in the dataset

In [ ]:
#lets have a look at the test dataset
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46477 entries, 0 to 46476
Data columns (total 41 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ID            46477 non-null  object 
 1   country_code  46477 non-null  int64  
 2   region        46477 non-null  int64  
 3   age           46348 non-null  float64
 4   FQ1           46477 non-null  int64  
 5   FQ2           20801 non-null  float64
 6   FQ3           19928 non-null  float64
 7   FQ4           46477 non-null  int64  
 8   FQ5           8913 non-null   float64
 9   FQ6           25878 non-null  float64
 10  FQ7           25917 non-null  float64
 11  FQ8           46477 non-null  int64  
 12  FQ9           46477 non-null  int64  
 13  FQ10          46477 non-null  int64  
 14  FQ11          35912 non-null  float64
 15  FQ12          46477 non-null  int64  
 16  FQ13          46477 non-null  int64  
 17  FQ14          46477 non-null  int64  
 18  FQ15          46477 non-nu

The test dataset has as total of 46477 rows with 41 features but missing the target features

**Lets have a look at the statistical analysis of the dataset as a whole**

In [ ]:
train.describe()

country_code         region  ...           FQ37         Target
count  108446.000000  108446.000000  ...  108446.000000  108446.000000
mean       68.544953       2.894242  ...       0.631457       0.273970
std        41.529264       2.286505  ...       0.482412       0.445996
min         0.000000      -1.000000  ...       0.000000       0.000000
25%        33.000000       1.000000  ...       0.000000       0.000000
50%        65.000000       3.000000  ...       1.000000       0.000000
75%       105.000000       4.000000  ...       1.000000       1.000000
max       143.000000       7.000000  ...       1.000000       1.000000

[8 rows x 41 columns]

1. Most of the max values in the feautres are 4 except for FQ19 so we should probably have a look at that later

2. Most of the features have the same standard deviation, minimum


In [ ]:
test.describe()

country_code        region  ...          FQ34          FQ37
count  46477.000000  46477.000000  ...  33136.000000  46477.000000
mean      68.364223      2.904017  ...      1.129466      0.631280
std       41.509763      2.282829  ...      0.354096      0.482463
min        0.000000     -1.000000  ...      1.000000      0.000000
25%       33.000000      1.000000  ...      1.000000      0.000000
50%       65.000000      3.000000  ...      1.000000      1.000000
75%      104.000000      4.000000  ...      1.000000      1.000000
max      143.000000      7.000000  ...      4.000000      1.000000

[8 rows x 40 columns]

**We would like to have a look at the columns and their datatypes**

Lets first have a look at the datatypes and their percentage missing values

In [ ]:
miss=train.isna().sum().sort_values(ascending=False)
missPerc=miss/len(train)*100
missPerc=pd.DataFrame(data=missPerc,columns=[['missPerc']])
missPerc[:15]

missPerc
FQ31  99.198680
FQ28  98.611290
FQ30  98.049721
FQ27  97.049223
FQ17  89.536728
FQ36  89.411320
FQ5   80.464932
FQ35  76.127289
FQ24  64.561164
FQ3   57.381554
FQ2   54.701879
FQ7   44.101212
FQ6   44.065249
FQ32  43.938919
FQ19  43.714844

The feature FQ31 has a missing percentage of close to 100% this is very sketchy and we hence have to figure out what is going on there. We also have to address the remaining train features that equally have high null values in their features

In [ ]:
#full list of all columns and their 
missPerc

missPerc
FQ31          99.198680
FQ28          98.611290
FQ30          98.049721
FQ27          97.049223
FQ17          89.536728
FQ36          89.411320
FQ5           80.464932
FQ35          76.127289
FQ24          64.561164
FQ3           57.381554
FQ2           54.701879
FQ7           44.101212
FQ6           44.065249
FQ32          43.938919
FQ19          43.714844
FQ34          29.317817
FQ20          22.756948
FQ21          22.716375
FQ11          22.656437
FQ29          22.623241
age            0.296922
FQ33           0.001844
country_code   0.000000
FQ1            0.000000
FQ4            0.000000
FQ9            0.000000
region         0.000000
FQ8            0.000000
Target         0.000000
FQ10           0.000000
FQ12           0.000000
FQ13           0.000000
FQ14           0.000000
FQ15           0.000000
FQ16           0.000000
FQ37           0.000000
FQ18           0.000000
FQ22           0.000000
FQ23           0.000000
FQ25           0.000000
FQ26           0.000000
ID             0.000000

In [ ]:
#lets have a look at the test data
miss=test.isna().sum().sort_values(ascending=False)
missPercT=miss/len(test)*100
missPercT=pd.DataFrame(data=missPerc,columns=[['missPerc']])
missPercT[:15]

missPerc
FQ31  99.216817
FQ28  98.642339
FQ30  98.115197
FQ27  96.895239
FQ17  89.504486
FQ36  89.457151
FQ5   80.822773
FQ35  76.220496
FQ24  64.358715
FQ3   57.122878
FQ2   55.244530
FQ6   44.320847
FQ7   44.236934
FQ32  44.058351
FQ19  43.800159

**Both have really high percentages of missing values lets concatinate the dataframes and have a look if each coreesponds to train and test values**

In [ ]:
missPD=pd.concat([missPerc,missPercT],axis=1)
missPD

missPerc   missPerc
FQ31          99.198680  99.216817
FQ28          98.611290  98.642339
FQ30          98.049721  98.115197
FQ27          97.049223  96.895239
FQ17          89.536728  89.504486
FQ36          89.411320  89.457151
FQ5           80.464932  80.822773
FQ35          76.127289  76.220496
FQ24          64.561164  64.358715
FQ3           57.381554  57.122878
FQ2           54.701879  55.244530
FQ7           44.101212  44.236934
FQ6           44.065249  44.320847
FQ32          43.938919  44.058351
FQ19          43.714844  43.800159
FQ34          29.317817  28.704521
FQ20          22.756948  22.497149
FQ21          22.716375  22.591820
FQ11          22.656437  22.731674
FQ29          22.623241  22.809131
age            0.296922   0.277557
FQ33           0.001844   0.000000
country_code   0.000000   0.000000
FQ1            0.000000   0.000000
FQ4            0.000000   0.000000
FQ9            0.000000   0.000000
region         0.000000   0.000000
FQ8            0.000000   0.000000
Target         0.000000        NaN
FQ10           0.000000   0.000000
FQ12           0.000000   0.000000
FQ13           0.000000   0.000000
FQ14           0.000000   0.000000
FQ15           0.000000   0.000000
FQ16           0.000000   0.000000
FQ37           0.000000   0.000000
FQ18           0.000000   0.000000
FQ22           0.000000   0.000000
FQ23           0.000000   0.000000
FQ25           0.000000   0.000000
FQ26           0.000000   0.000000
ID             0.000000   0.000000

**This gives a rough idea of how the training and testing split happened**
Looks like they created a threshold to ensure that both train and test had almost similar null values.Hence it wasn't random at all





**We now go through each column to understand it better we begin with all numeical columns**

## Numerical Columns

In [ ]:
#create a list of all numeric columns
num=train.select_dtypes(exclude='object')

In [ ]:
num.columns

Index(['country_code', 'region', 'age', 'FQ1', 'FQ2', 'FQ3', 'FQ4', 'FQ5',
       'FQ6', 'FQ7', 'FQ8', 'FQ9', 'FQ10', 'FQ11', 'FQ12', 'FQ13', 'FQ14',
       'FQ15', 'FQ16', 'FQ17', 'FQ18', 'FQ19', 'FQ20', 'FQ21', 'FQ22', 'FQ23',
       'FQ24', 'FQ35', 'FQ36', 'FQ25', 'FQ26', 'FQ27', 'FQ28', 'FQ29', 'FQ30',
       'FQ31', 'FQ32', 'FQ33', 'FQ34', 'FQ37', 'Target'],
      dtype='object')

In [ ]:
len(num)

108446

In [ ]:
""""
 we create a function that does the following
 1. calculates the quantile statistcis of all the columns
 2.calculates the descripritve stats of the columns
 3.creates a histogram of the columns 

"""

#we can creat a dictionary that stores each of the columns and their respective values hence creating a dataframe with the info


def quantileStats(rows):
  stats={}
  for row in rows.columns:
    stats[row]=rows[row].describe()
    statsPD=pd.DataFrame(stats)
  return statsPD


In [ ]:
quantileStats(num)

country_code         region  ...           FQ37         Target
count  108446.000000  108446.000000  ...  108446.000000  108446.000000
mean       68.544953       2.894242  ...       0.631457       0.273970
std        41.529264       2.286505  ...       0.482412       0.445996
min         0.000000      -1.000000  ...       0.000000       0.000000
25%        33.000000       1.000000  ...       0.000000       0.000000
50%        65.000000       3.000000  ...       1.000000       0.000000
75%       105.000000       4.000000  ...       1.000000       1.000000
max       143.000000       7.000000  ...       1.000000       1.000000

[8 rows x 41 columns]

In [ ]:
def DescriptiveStats(rows):
  dStats={}
  for row in rows.columns:
    dStats[row]={'skew':stats.skew(rows[row]),'kurtosis':stats.kurtosis(rows[row])}
    dStatsPD=pd.DataFrame(dStats)

  return dStatsPD

In [ ]:
stats.skew(train.FQ1)

0.2323922243314188

In [ ]:
descriptiveDF=DescriptiveStats(num)
descriptiveDF=descriptiveDF.T
descriptiveDF

skew   kurtosis
country_code  0.115108  -1.174547
region        0.213910  -1.125476
age                NaN        NaN
FQ1           0.232392  -0.288118
FQ2                NaN        NaN
FQ3                NaN        NaN
FQ4          -0.558372   2.532145
FQ5                NaN        NaN
FQ6                NaN        NaN
FQ7                NaN        NaN
FQ8          -0.815478   5.030799
FQ9          -0.477346   2.210386
FQ10         -0.375672   1.742524
FQ11               NaN        NaN
FQ12         -1.121865   5.158917
FQ13         -1.239373   7.926392
FQ14         -1.457873  13.406937
FQ15         -1.054843   6.147937
FQ16         -0.674177   1.473942
FQ17               NaN        NaN
FQ18          0.915058   0.670290
FQ19               NaN        NaN
FQ20               NaN        NaN
FQ21               NaN        NaN
FQ22          0.651703  -0.239773
FQ23         -0.161807  -0.115717
FQ24               NaN        NaN
FQ35               NaN        NaN
FQ36               NaN        NaN
FQ25         -1.010016   4.019321
FQ26         -1.068656   5.077993
FQ27               NaN        NaN
FQ28               NaN        NaN
FQ29               NaN        NaN
FQ30               NaN        NaN
FQ31               NaN        NaN
FQ32               NaN        NaN
FQ33               NaN        NaN
FQ34               NaN        NaN
FQ37         -0.545002  -1.702973
Target        1.013599  -0.972617

* **If the skewness is between -0.5 and 0.5, the data are fairly symmetrical**
* **If the skewness is between -1 and — 0.5 or between 0.5 and 1, the data are moderately skewed**
* **If the skewness is less than -1 or greater than 1, the data are highly skewed**

In [ ]:
#lets get the higly skewed columns

descriptiveDF[(descriptiveDF['skew']>1)| (descriptiveDF['skew']<-1) ]


skew   kurtosis
FQ12   -1.121865   5.158917
FQ13   -1.239373   7.926392
FQ14   -1.457873  13.406937
FQ15   -1.054843   6.147937
FQ25   -1.010016   4.019321
FQ26   -1.068656   5.077993
Target  1.013599  -0.972617

* If the absolute value of skew < 0.5 then very symmetric.
* If the absolute value of skew is in between 0.5 and 1 then slightly skewed
* If the absolute value of skew is greater than 1 then very skewed.

In [ ]:
descriptiveDF[(descriptiveDF['kurtosis']>1)| (descriptiveDF['kurtosis']<-1) ]

skew   kurtosis
country_code  0.115108  -1.174547
region        0.213910  -1.125476
FQ4          -0.558372   2.532145
FQ8          -0.815478   5.030799
FQ9          -0.477346   2.210386
FQ10         -0.375672   1.742524
FQ12         -1.121865   5.158917
FQ13         -1.239373   7.926392
FQ14         -1.457873  13.406937
FQ15         -1.054843   6.147937
FQ16         -0.674177   1.473942
FQ25         -1.010016   4.019321
FQ26         -1.068656   5.077993
FQ37         -0.545002  -1.702973

 we will have to do something about the columns with excess kurtosis

In [ ]:
train[train.duplicated()]

Empty DataFrame
Columns: [ID, country_code, region, age, FQ1, FQ2, FQ3, FQ4, FQ5, FQ6, FQ7, FQ8, FQ9, FQ10, FQ11, FQ12, FQ13, FQ14, FQ15, FQ16, FQ17, FQ18, FQ19, FQ20, FQ21, FQ22, FQ23, FQ24, FQ35, FQ36, FQ25, FQ26, FQ27, FQ28, FQ29, FQ30, FQ31, FQ32, FQ33, FQ34, FQ37, Target]
Index: []

In [ ]:
test[test.duplicated()]

Empty DataFrame
Columns: [ID, country_code, region, age, FQ1, FQ2, FQ3, FQ4, FQ5, FQ6, FQ7, FQ8, FQ9, FQ10, FQ11, FQ12, FQ13, FQ14, FQ15, FQ16, FQ17, FQ18, FQ19, FQ20, FQ21, FQ22, FQ23, FQ24, FQ35, FQ36, FQ25, FQ26, FQ27, FQ28, FQ29, FQ30, FQ31, FQ32, FQ33, FQ34, FQ37]
Index: []

In [ ]:
sns.pairplot(num,hue='Target')

KeyboardInterrupt: ignored

Error in callback <function install_repl_displayhook.<locals>.post_execute at 0x7f20f5ca6dd0> (for post_execute):


KeyboardInterrupt: ignored

Error in callback <function flush_figures at 0x7f20f5caf440> (for post_execute):


KeyboardInterrupt: ignored

## univariate on Categorical columns

In [ ]:
cat=train.select_dtypes(include='object')

In [ ]:
cat

ID
0       ID_000J8GTZ
1       ID_000QLXZM
2       ID_001728I2
3       ID_001R7IDN
4       ID_0029QKF8
...             ...
108441  ID_ZZYY7RTO
108442  ID_ZZYZTTC6
108443  ID_ZZZ3OW3S
108444  ID_ZZZLDXE8
108445  ID_ZZZMYW1F

[108446 rows x 1 columns]

**we probably wont need this col so yup no need to keep it around**

# Feature selection

In [ ]:
stats.entropy(num.age)

nan

In [ ]:
#creating a column that generates the entropy of the numeric categorical features

def entNumeric(rows):
  ent={}
  for row in rows.columns:
    ent[row]=stats.entropy(rows[row])

    
    entPD=pd.DataFrame(ent,index=['Entropy'])
  return entPD



In [ ]:
entTable=entNumeric(num).T

In [ ]:
entTable

Entropy
country_code  11.383478
region             -inf
age                 NaN
FQ1           11.535821
FQ2                 NaN
FQ3                 NaN
FQ4           11.562742
FQ5                 NaN
FQ6                 NaN
FQ7                 NaN
FQ8           11.570931
FQ9           11.560010
FQ10          11.556611
FQ11                NaN
FQ12          11.572790
FQ13          11.577375
FQ14          11.582273
FQ15          11.574111
FQ16          11.559935
FQ17                NaN
FQ18          11.521657
FQ19                NaN
FQ20                NaN
FQ21                NaN
FQ22          11.530105
FQ23          11.544943
FQ24                NaN
FQ35                NaN
FQ36                NaN
FQ25          11.570399
FQ26          11.572795
FQ27                NaN
FQ28                NaN
FQ29                NaN
FQ30                NaN
FQ31                NaN
FQ32                NaN
FQ33                NaN
FQ34                NaN
FQ37          11.134282
Target        10.299273

## Missing values

In [ ]:
#lets fill in the missing values with zero initially

train.fillna(0,inplace=True)
test.fillna(0,inplace=True)

# Model Training

In [ ]:
# we train a simple Random Forest Regressor
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import auc
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split,GridSearchCV,RandomizedSearchCV,StratifiedKFold



In [ ]:
X=train.drop(labels='Target',axis=1)
y=train['Target']

In [ ]:
X.drop(labels='ID',axis=1,inplace=True)

In [ ]:
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=123,test_size=0.3)
RF=RandomForestClassifier(n_estimators=100,n_jobs=-1,verbose=True)
SKF=StratifiedKFold(n_splits=5)



Lets create a base model then tune it later

In [ ]:
RF.fit(X_train,y_train)


[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  46 tasks      | elapsed:    6.2s
[Parallel(n_jobs=-1)]: Done 100 out of 100 | elapsed:   13.4s finished


RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=-1, oob_score=False, random_state=None,
                       verbose=True, warm_start=False)

In [ ]:
RF.score(X_train,y_train)

[Parallel(n_jobs=2)]: Using backend ThreadingBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  46 tasks      | elapsed:    1.0s
[Parallel(n_jobs=2)]: Done 100 out of 100 | elapsed:    2.2s finished


1.0

In [ ]:
RF.score(X_test,y_test)

[Parallel(n_jobs=2)]: Using backend ThreadingBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  46 tasks      | elapsed:    0.5s
[Parallel(n_jobs=2)]: Done 100 out of 100 | elapsed:    1.0s finished


0.7236122210610438

In [ ]:
#lets get predicitions
predictions=RF.predict(X_test)

[Parallel(n_jobs=2)]: Using backend ThreadingBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  46 tasks      | elapsed:    0.4s
[Parallel(n_jobs=2)]: Done 100 out of 100 | elapsed:    1.0s finished


In [ ]:
from sklearn.metrics import roc_auc_score

In [ ]:
roc_auc_score(y_test,predictions)

0.5001294289309259

In [ ]:
IDs=test['ID']
testF=test.drop(labels='ID',axis=1)

In [ ]:
testPredictions=RF.predict(testF)

[Parallel(n_jobs=2)]: Using backend ThreadingBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  46 tasks      | elapsed:    0.6s
[Parallel(n_jobs=2)]: Done 100 out of 100 | elapsed:    1.3s finished


In [ ]:
testPredictions

array([0, 0, 0, ..., 0, 0, 0])

In [ ]:
testPD=pd.DataFrame(data=[IDs,testPredictions])

In [ ]:
testPD=testPD.T

In [ ]:
testPD

ID Unnamed 0
0      ID_000YI58E         0
1      ID_001SP4JF         0
2      ID_001VOF6S         0
3      ID_0030LULG         0
4      ID_0037PZ3R         0
...            ...       ...
46472  ID_ZZYOTVBJ         0
46473  ID_ZZYSX122         0
46474  ID_ZZYXQDSD         0
46475  ID_ZZZH9SS4         0
46476  ID_ZZZTLL7U         0

[46477 rows x 2 columns]

In [ ]:
testPD.columns=['ID','Target']

In [ ]:
testPD

ID Target
0      ID_000YI58E      0
1      ID_001SP4JF      0
2      ID_001VOF6S      0
3      ID_0030LULG      0
4      ID_0037PZ3R      0
...            ...    ...
46472  ID_ZZYOTVBJ      0
46473  ID_ZZYSX122      0
46474  ID_ZZYXQDSD      0
46475  ID_ZZZH9SS4      0
46476  ID_ZZZTLL7U      0

[46477 rows x 2 columns]

In [ ]:
testPD.Target.value_counts()

0    46469
1        8
Name: Target, dtype: int64

In [ ]:
testPD.to_csv('firstSub.csv',index=False)